Add sandbox

markusjura opened the pull request:
This PR is adding the sandbox to the conductr-cli.

markusjura commented:
This PR is still work in progress. 

**Done**
- CLI parser for `sandbox run` and `sandbox stop` (beside some cosmetics)
- Implementation of `sandbox run`

*ToDo*
- Implementation of `sandbox stop`
- Implementation of `sandbox debug` (mostly reuse of `sandbox run` functions)
- CLI parser for `sandbox debug (reuse of `sandbox run` parser)
- Tests

markusjura commented:
@huntc: Can you do a first review?

huntc commented:
Sure thing. Excited.

huntc commented on conductr_cli/sandbox.py:
```diff
@@ -0,0 +1,115 @@
+def add_default_arguments(sub_parser):
+    sub_parser.add_argument('-i', '--ip',
+                            default=DEFAULT_IP,
+                            help='ConductR IP address. Defaults to $CONDUCTR_IP or 127.0.0.1.')
```
Not sure that I understand why you'd specify an IP for the sandbox...

markusjura replied:
We need that to expose the {ip}:{port} when starting the sandbox.

Generally I was thinking about implementing the same `default ip` logic into the `conductr-cli` as we have it in `sbt-conductr`:
- Try to get ip from `docker-machine`, else
  - Try to get ip from `boot2docker`, else
    - Try to get ip from `hostname`, else
      - 127.0.0.1

WDYT?

huntc commented on conductr_cli/sandbox.py:
```diff
@@ -0,0 +1,115 @@
+    run_parser.add_argument('-v', '--image-version',
+                            default='1.0.11',
+                            help='Version of the Docker image to use.\n'
+                                 'Defaults to 1.0.11.\n'
```
1.0.12 is released, and the Docker image for it should be available in less than 24 hours. Meanwhile, I suggest changing this now, and tagging a 1.0.12 in your local registry.

markusjura replied:
Sure, just wanted to test that yesterday evening..

huntc commented on conductr_cli/sandbox.py:
```diff
@@ -0,0 +1,115 @@
+
+
+if __name__ == '__main__':
+    run()
```
Reallly nice

huntc commented on conductr_cli/sandbox_start.py:
```diff
@@ -0,0 +1,159 @@
+                   9005, # ConductR controlServer
+                   9006, # ConductR bundleStreamServer
+                   9200, # conductr-elasticsearch bundle
+                   9999} # visualizer bundle
```
sp before }

markusjura replied:
Done

huntc commented on conductr_cli/sandbox_start.py:
```diff
@@ -0,0 +1,159 @@
+
+#############################
+###       Constants       ###
+#############################
```
Have we been using this commenting convention elsewhere? If not then either we apply globally, or not at all.

huntc commented on conductr_cli/sandbox_start.py:
```diff
@@ -0,0 +1,159 @@
+
+def debug(args):
+    """`sandbox debug` command"""
+    print("TODO: Implement")
```
I'm starting to wonder about the utility of debugging outside of an build environment such as sbt. WDYT?

huntc commented on conductr_cli/sandbox_start.py:
```diff
@@ -0,0 +1,159 @@
+
+# Stop nodes
+def stop_nodes():
+    '''TODO: Implement'''
```
Just wondering if you should seek any inspiration from the original run_conductrs python script we had - perhaps not, but just thought I'd remind you of it.

huntc commented on conductr_cli/terminal.py:
```diff
@@ -0,0 +1,24 @@
+def docker_run(optional_args, image, positional_args):
+    command = ['docker', 'run'] + optional_args + [image] + positional_args
+    print('docker command: {}'.format(' '.join(command)))
+    return subprocess.call(['docker', 'run'] + optional_args + [image] + positional_args)
```
Is console a better name than terminal?

markusjura commented:
**Update**
- Review comments have been resolved / implemented
- Logic to automatically resolve host IP address with `docker-machine`, `boot2docker`, `hostname` has been implemented
- `sandbox` parser now checks if the `image_version` has been set. Otherwise the program exits with a message that the version need to be obtained from typesafe.com (same message as in `sbt-conductr-sandbox`)
- `sandbox` parser has been optimized
- `sandbox debug` command has been implemented (disabled for now, as discussed with @huntc)
- `sandbox stop` command has been implemented
- `conductr` and `sandbox` parser have been aligned (same help output)
- README.rst has been updated
- Tests are still missing

In summary the new `sandbox` command fully works with the conductr image `conductr/conductr` and the version `1.0.12`, but no tests have been written yet.

markusjura commented:
Implementation is done, smoke tests on linux, OS X and Windows have been successful.  Only unit tests are still missing.

**Today's progress**
- Added latest python3 version to test config (currently Python 3.5.0)
- Fixed `conduct` command
- Fixed sandbox port mapping logic
- Improved `sandbox` error handling for linux
- Fixed the `sandbox --feature` option
- Smoke tests on OS X and linux (sandbox run, stop, run with features, conduct info, load, run, stop)

huntc commented:
Tests on CI now pass also. Awesome. I can't wait to try this.

edwardcallahan commented:
Looking good. Tested visualization, multi container, service lookup, general bundle usage on Linux. Nice work, Markus.

huntc commented:
If we attempt to run a `conduct` command and there is no ConductR running, is it reasonable to also suggest starting the sandbox (bearing in mind we only do so when --ip points to a VM)? Presently:

```
Christophers-MacBook-Pro:typesafe-conductr-cli huntc$ python3 -m conductr_cli.conduct services
ERROR: Unable to contact ConductR.
ERROR: Reason: HTTPConnectionPool(host='192.168.59.103', port=9005): Max retries exceeded with url: /bundles (Caused by NewConnectionError('<requests.packages.urllib3.connection.HTTPConnection object at 0x10a868e80>: Failed to establish a new connection: [Errno 61] Connection refused',))
ERROR: Make sure it can be accessed at http://192.168.59.103:9005/bundles
```

huntc commented:
It'd be nice to see an example of setting the roles in the help text.

huntc commented:
Should we align sbt-sandbox's --with-feature with the sandbox here i.e. --feature for each feature...?

markusjura commented:
Created an issue to align the arguments: https://github.com/typesafehub/sbt-conductr-sandbox/issues/63

markusjura commented:
> It'd be nice to see an example of setting the roles in the help text.

Yes, will add this.

markusjura commented:
Introduced a new argument `--local-connection` which is not displayed in the help. This argument is true if the `--ip` argument is not getting overriden in the `conduct` command. For the `sandbox` commands `--local-connection` is always True.

If `--local-connection` is True then the error message is different if ConductR can't be reached:

```
mj@MacBook-Pro-2:~/workspace/conductr-cli$ python3 -m conductr_cli.conduct info
ERROR: Unable to contact ConductR.
ERROR: Reason: HTTPConnectionPool(host='192.168.99.100', port=9005): Max retries exceeded with url: /bundles (Caused by NewConnectionError('<requests.packages.urllib3.connection.HTTPConnection object at 0x1101e75c0>: Failed to establish a new connection: [Errno 61] Connection refused',))
ERROR: Start the ConductR sandbox with: sandbox run IMAGE_VERSION
```

The last line changed accordingly pointing the user towards how to start the sandbox.

huntc commented:
When would a user need to use the local connection arg?

markusjura commented:
Never, it is basically a hidden argument which the user does not see in the help and which we will not document. We need this flag to check during error handling if the connection is a local or remote connection, i.e. if the user has specified the `--ip` arg. During error handling we can't check this based on the `--ip` argument anymore because the value is set to an ip address. Therefore we need an additional argument `--local-connection`.

huntc commented:
Feels then like we are overloading the use of args when what we are trying to do is set an internal flag... WDYT?

markusjura commented:
I think it is fine. In the `argparse` library there is a notation to `suppress` things which we use in this case:

```
sub_parser.add_argument('--local-connection',
  default=True,
  help=argparse.SUPPRESS)
```

So it seems to be OK. Also if we would not put this flag into the args object then it would make things much more complicated. We couldn't use `parser.set_defaults()` anymore and at the end call `args.func(args)`.

markusjura commented:
Will push the tests on Sunday and then merge this PR on Sunday as well.

huntc commented:
Thanks Markus! Happy to review the tests also. Btw the tests are presently failing on this PR.

markusjura commented:
Tests has been added.

@huntc: Can you please do a final review?

huntc commented:
Commits need to be squashed.

huntc commented on conductr_cli/sandbox.py:
```diff
@@ -0,0 +1,133 @@
+                            type=int,
+                            default=5005,
+                            help='Debug port to be made public by each of the ConductR containers. Defaults to 5005.',
+                            metavar='')
```
Is this arg still required?

markusjura replied:
No, it's not. Good catch.

huntc commented:
Other than https://github.com/typesafehub/conductr-cli/pull/73#issuecomment-155232196 and https://github.com/typesafehub/conductr-cli/pull/73#discussion_r44351682, LGTM. Please address those two things and then merge/release.

Thanks for the awesomeness!

markusjura committed:
Add sandbox

markusjura commented:
Mistakenly renamed the branch and therefor PR got closed.

markusjura commented:
Commits have been squashed.
